README.md
# Semantic Data Taking Web-app

This app is a Flask-based dynamic website that uses an RDF graph database to
store data. Initially, the website is for manually entering data into HTML
forms. The entered data can then be viewed in tables.

## Run locally with Docker

1. Make directories:
   ```bash
   mkdir -p certs
   ```
2. Generate a self-signed certificate:
   ```bash
   openssl req -x509 -newkey rsa:4096 -keyout certs/key.pem -out certs/cert.pem -sha256 -days 365 -nodes -subj '/CN=semweb.localhost'
   ```
4. Start the services:
   ```bash
   docker compose up
   ```
3. Initialize the database by running:
   ```bash
   docker compose run semantic-app pipenv run python manage.py init http://db-server:3030/semweb
   ```
   and then restart the semantic-app:
   ```bash
   docker compose restart semantic-app
   ```
5. Generate a userfile with
   ```bash
   pipenv sync
   pipenv run flask-simple-login-gen-user-file-line > userfile.txt
   ```
   then copy it to the container with:
   ```bash
   docker cp userfile.txt <containerid>:/app/users/userfile.txt
   ```
   where you can find the containerid with `docker container ls` as the line
   with "semantic-data-taking-webapp-semantic-app".

The app should be visible at https://semweb.localhost:8080

## Run Test Server

First, make sure BerkeleyDB is installed on your system:

```bash
sudo apt install libdb-dev
```

Set things up:

```bash
pipenv install
pipenv shell
```

Create a user:

```bash
flask-simple-login-gen-user-file-line > userfile.txt
```

Initialize the database:

```bash
python db.py --init berkeleydb
```

Run the server:

```bash
python web.py
```

Access the site at https://semweb.localhost:5000

The test site uses a dummy SSL certificate that updates on every server
restart, so you will have to allow that in your browser after every update.

## Backup the DB

Click the "Download All Data in RDF TriG" button on the main web page.

**OR**

On the docker server, download this git repository and cd into the directory, and run

```bash
pipenv sync
pipenv run python manage.py dump http://localhost:3030/semweb semweb_backup.trig
```

## Import a backup into the DB

1. Setup a new docker DB server and don't initialize the DB

2. Start the docker server with `docker compose -f docker-compose-prod.yml up` or similar

2. Download this git repository and cd into the directory, then run
   ```bash
   pipenv sync
   pipenv run python manage.py write --content-type="application/trig" http://localhost:3030/semweb semweb_backup.trig
   ```
   and replace "semweb_backup.trig" with your backup file

## Setting up for development:

```bash
pre-commit install
pre-commit install --hook-type commit-msg
```

You may have to do run this as well to get pre-commit to work with your python:

```bash
pipenv run pip install -U pre-commit black
```

and this before every commit:

```bash
pipenv run pre-commit run --all
```

## Making a Release

This project uses SemVer, automatically via Uplift. On a tagged commit, a release may be made by clicking on the "releases" section of the GitHub repository page and then the "Draft a new release" button. Once a release is published, a Docker image will automatically be generated.

## Example Scenario

A simple example of usage would be to collect data on the fuel economy of a
car. When filling up a car, the user would enter the odometer reading and
amount of fuel put into the car.

## Schema Design

Custom classes are prefixed with `sdtw`.

- Mainly based on the [Semantic Sensor Network Ontology](https://www.w3.org/TR/vocab-ssn/)
  - In this case the data-entering user would be of type `sosa:Sensor` and also either a `foaf:Person` or `vcard:Individual`
  - The particular car would be of type `sosa:FeatureOfInterest`
  - The odometer reading would be a `sosa:ObservableProperty`
    - Each record of the odometer reading would be a `sosa:Observation` with the value `sosa:hasResult` and recording time `sosa:resultTime` (a `xsd:dateTime`)
      - I'd like to include units, so the `sosa:Result` could also be of type `qudt:Quantity` from [QUDT](https://www.qudt.org/) and be given a `qudt:unit` and the actual value would be stored as a `qudt:value`
    - Each `sosa:ObservableProperty` would `sdtw:hasQuantityKind` and `sdtw:hasUnit` the `qudt:QuantityKind` and `qudt:Unit` the `sosa:Observable` should be given in. These serve as meta-data for building the forms and tables.
  - The amount of fuel put in the car would be another `sosa:ObservableProperty`
    - Observations would be made similar to with the odometer reading
  - What links together one particular odometer reading and amount of fuel is the same `ssn:Stimulus`.
    - The stimulus would `sdtw:hasTime` the same timestamp as the observations have as resultTime

So the only custom Ontological pieces would be:

- `sdtw:hasQuantityKind` mapping from `sosa:ObservableProperty` to `qudt:QuantityKind`
- `sdtw:hasUnit` mapping from `sosa:ObservableProperty` to `qudt:Unit`
- `sdtw:hasTime` mapping from `ssn:Stimulus` to `xsd:dataTime` (or `sosa:resultTime`?)

## User Interface Design

*Assuming the schema (metadata including feature of interest and observable properties) for a data taking scenario is already in the database for now*

- Top level menu chooses the scenario or thing to enter/view data about. This would be a list of buttons for each `sosa:FeatureOfInterest`
  - A menu to choose data entry or view
    - Data entry:
      - A form field is presented for each `sosa:ObservableProperty` listed with the `qudt:QuantityKind` and `qudt:Unit` it should be given in
      - The submission time is recorded as the resultTime
    - Data view:
      - A table
        - Columns: the resultTime and one for each `sosa:ObservableProperty`
        - Rows: One for each `ssn:Stimulus`

## Database Key Design

Assuming all of this is present in `ex:`.

- `sosa:Sensor`/`foaf:Person`: `ex:users/<username>`
- `sosa:FeatureOfInterest`: `ex:features/<featurename>`
- `sosa:ObservableProperty`: `ex:properties/<featurename>/<propertyname>`
- `ssn:Stimulus`: `ex:stimuli/<featurename>/<resultTime>`
- `sosa:Observation`: `ex:observations/<featurename>/<propertyname>/<resultTime>`
- `sosa:Result`: `ex:results/<featurename>/<propertyname>/<resultTime>`

Everything else would be anonymous.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## [v1.11.5](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.11.5) - 2023-07-05

- [`d809efd`](https://github.com/jhugon/semantic-data-taking-webapp/commit/d809efd0cd34d8efc8a73d8339e37b0242180668) fix: Load and treat Currency different from other quantityKind

## [v1.11.4](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.11.4) - 2023-07-05

- [`dacf5e6`](https://github.com/jhugon/semantic-data-taking-webapp/commit/dacf5e65965ff9479fbd3c376595cc98a5624e45) fix: only load quantity kinds english labels

## [v1.11.3](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.11.3) - 2023-07-05

- [`8a4f7ab`](https://github.com/jhugon/semantic-data-taking-webapp/commit/8a4f7ab4104edcb3a66165af8908bac796253b04) fix: Made docker-compose files consistent

## [v1.11.2](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.11.2) - 2023-07-04

- [`98861ae`](https://github.com/jhugon/semantic-data-taking-webapp/commit/98861aec287f84cae03a813f91d57b958e06dceb) fix: docker-compose-prod.yml works in prod

## [v1.11.1](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.11.1) - 2023-07-03

- [`9a358c4`](https://github.com/jhugon/semantic-data-taking-webapp/commit/9a358c4e71bd9f7c1b886b3fd6e91caa1bbdf4d3) Publish docker image on release (#59)

## [v1.11.0](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.11.0) - 2023-07-01

- [`4229420`](https://github.com/jhugon/semantic-data-taking-webapp/commit/4229420252ec9a389a3c0643f29f8464e343639f) feat: Update docker.yml

## [v1.10.0](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.10.0) - 2023-06-29

- [`c0182ae`](https://github.com/jhugon/semantic-data-taking-webapp/commit/c0182ae4a62ba119ff166535e59ce118a92d1b86) Restore DB (#55)

## [v1.9.0](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.9.0) - 2023-05-04

- [`1cf0bb1`](https://github.com/jhugon/semantic-data-taking-webapp/commit/1cf0bb148986da83e0ca009677d8fd30a2cece3a) Jena fuseki docker image (#53)

## [v1.8.1](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.8.1) - 2023-05-04

- [`c17ae7f`](https://github.com/jhugon/semantic-data-taking-webapp/commit/c17ae7fe9e882af4ff400250f7f75c70f9e52434) fix(deps): Security upgrade to Flask

## [v1.8.0](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.8.0) - 2023-04-27

- [`fea6a10`](https://github.com/jhugon/semantic-data-taking-webapp/commit/fea6a104c25feab07e13c04f6618d1f609f8fbc2) feat: added docker-compose-prod.yml

## [v1.7.0](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.7.0) - 2023-04-26

- [`a2171a8`](https://github.com/jhugon/semantic-data-taking-webapp/commit/a2171a8f3db4822781375cd269f0e43539a5c2ff) feat(docker): now have to run manually

## [v1.6.1](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.6.1) - 2023-04-26

- [`3891602`](https://github.com/jhugon/semantic-data-taking-webapp/commit/3891602b774093fb87aad22bda9ca91d54f1ce70) fix: trying to get uplift to trigger github action

## [v1.6.0](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.6.0) - 2023-04-26

- [`adf3508`](https://github.com/jhugon/semantic-data-taking-webapp/commit/adf3508e6245921a0d0966a40e20d6fd2093575d) ci: re-enable docker workflow on tag push
- [`c3d589d`](https://github.com/jhugon/semantic-data-taking-webapp/commit/c3d589d6ac0af7987b4eb4d7727cf8b4c49d0c0a) feat: update Flask version to ~2.3.0

## [v1.5.7](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.5.7) - 2023-04-19

- [`84e19bd`](https://github.com/jhugon/semantic-data-taking-webapp/commit/84e19bdc2b0e8a1ac3fff8b5f4b7a25635c9ad6c) fix: trying to get uplift to trigger action

## [v1.5.6](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.5.6) - 2023-04-18

- [`9809d67`](https://github.com/jhugon/semantic-data-taking-webapp/commit/9809d67b463b30c2acd7d3f7b4c745667848c7ed) fix(build): fixing the pipenv version to add repetability
- [`f2417b5`](https://github.com/jhugon/semantic-data-taking-webapp/commit/f2417b5d48a3c20e34389726d38c8b9fae608c52) ci(docker): stop making releases
- [`7a00e2e`](https://github.com/jhugon/semantic-data-taking-webapp/commit/7a00e2eb8c9afc195d96bcc0370873c68396db93) ci(docker): checkout submodules
- [`ab9c751`](https://github.com/jhugon/semantic-data-taking-webapp/commit/ab9c751ed2264414842710c3e370b508722a0dc9) ci: try to make docker workflow manually triggerable

## [v1.5.5](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.5.5) - 2023-03-29

- [`0682a74`](https://github.com/jhugon/semantic-data-taking-webapp/commit/0682a743efe65e3bdaf7cda993e0661b8f0e9ae4) fix: uplift workflow now uses personal access token

## [v1.5.4](https://github.com/jhugon/semantic-data-taking-webapp/releases/tag/v1.5.4) - 2023-03-29

- [`7937f82`](https://github.com/jhugon/semantic-data-taking-webapp/commit/7937f828013e535594c0f7df940a15ee94b2723a) fix: trying to fix docker CI, trigger on tags now
- [`3b37552`](https://github.com/jhugon/semantic-data-taking-webapp/commit/3b37552f7d79a843505fb72b1463f5f1ca4b2723) ci: Still trying to get docker workflow to work
- [`0b3f83a`](https://github.com/jhugon/semantic-data-taking-webapp/commit/0b3f83a41424ae5962d5fb8313141516be36a375) ci: try to fix by adding checkout action to docker workflow
- [`6ec6ed7`](https://github.com/jhugon/semantic-data-taking-webapp/commit/6ec6ed7dbab4389d77a7e2bc0de970022c99e4d0) ci: adding docker image build workflow
- [`423d726`](https://github.com/jhugon/semantic-data-taking-webapp/commit/423d726a0d4ed54ece8443d12aed05da99973e6b) build: updated Pipfile and lock


manage.py
from db import DBInterface, graph_store_post, graph_store_get
from enum import Enum
from pathlib import Path
from urllib.parse import urlparse
import logging
import typer
from rdflib import Graph
from rdflib.compare import to_isomorphic, graph_diff


class DBType(str, Enum):
    BerkelyDB = "BerkelyDB"
    SPARQLUpdateStore = "SPARQLUpdateStore"


app = typer.Typer()


@app.command()
def init(dbpath: str, dbtype: DBType = DBType.SPARQLUpdateStore):
    """
    Initializes the RDF database at DBPATH with standard quantities and units.
    """
    DBInterface.initialize_store(dbpath, dbtype.value)


@app.command()
def dump(
    dbpath: str,
    outfile: Path,
    content_type: str = "application/trig",
):
    """
    Dumps the contents of the RDF database at DBPATH to OUTFILE.
    """
    data = graph_store_get(dbpath, content_type=content_type)
    with outfile.open("w") as outf:
        outf.write(data)


@app.command()
def write(
    dbpath: str,
    infile: Path,
    data_uri_base: Path = None,
    content_type: str = "application/trig",
):
    """
    Writes the contents of INFILE into the RDF database at DBPATH.
    """
    with infile.open() as inf:
        graph_store_post(
            dbpath, inf.read(), graph_uri=data_uri_base, content_type=content_type
        )


@app.command()
def compare(graph1: Path, graph2: Path):
    """
    Compare the RDF graph, GRAPH1 to the RDF graph, GRAPH2. They must both be local files
    """

    graph1loc = graph1
    graph2loc = graph2

    graph1 = Graph()
    graph1.parse(graph1loc)
    graph2 = Graph()
    graph2.parse(graph2loc)
    graph1iso = to_isomorphic(graph1)
    graph2iso = to_isomorphic(graph2)
    if graph1iso == graph2iso:
        print("Graphs are isomorphic")
    else:
        print("Graphs are not isomorphic")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    app()


web.py
from flask import Flask
from flask import request, Response
from flask import render_template, redirect, url_for
from werkzeug.middleware.proxy_fix import ProxyFix
from flask_login import LoginManager, current_user
from flask_talisman import Talisman

from db import DBInterface, DataValidationError, GetSubjectError, QUDT, GEO
from flask_simple_login import (
    auth,
    User,
    login_required,
)


import urllib
from datetime import datetime, timedelta
import os
import re
import os.path
import logging
import sys


def create_app():
    """
    App factory
    """

    logging.basicConfig(
        format="%(levelname)s %(name)s: %(message)s",
        # format="%(asctime)s %(levelname)s %(name)s L%(lineno)s: %(message)s",
        level=logging.INFO,
    )

    app = Flask(__name__)

    #### Configuration ###################

    app.config["DB_STORE_PATH"] = "http://localhost:3030/semweb"
    app.config["DB_STORE_TYPE"] = "SPARQLUpdateStore"
    app.config["DB_DATA_URI_BASE"] = "http://data-webapp.hugonlabs.com/test1/"
    app.config["DB_USERS_URI_BASE"] = os.path.join(
        app.config["DB_DATA_URI_BASE"], "users/"
    )
    app.config["LOGIN_USER_FILE_PATH"] = "userfile.txt"
    app.config["PROXY_FORWARDING"] = True

    app.config["SERVER_NAME"] = "semweb.localhost"
    app.config["SECRET_KEY"] = b"dummy"

    app.config["PREFERRED_URL_SCHEME"] = "https"
    app.config["SESSION_PROTECTION"] = "strong"
    app.config["SESSION_COOKIE_SECURE"] = True
    app.config["SESSION_COOKIE_SAMESITE"] = "Strict"

    app.config["LOGIN_REMEMBER"] = True
    app.config["REMEMBER_COOKIE_DURATION"] = timedelta(days=7)
    app.config["REMEMBER_COOKIE_SECURE"] = True
    app.config["REMEMBER_COOKIE_REFRESH_EACH_REQUEST"] = True

    ## override above with contents of file in this environment variable:
    try:
        app.config.from_envvar("SEMWEB_SETTINGS")
    except RuntimeError as e:
        app.logger.warning(e)
    ## override above with environment variables prefixed with "FLASK_" e.g. "FLASK_SERVER_NAME"
    app.config.from_prefixed_env()
    for k in os.environ:
        rematch = re.match(r"^FLASK_(.*SECRET.*)", k)
        if rematch:
            app.logger.info(
                f"Configuration variable {rematch.group(1)} set from environment variable"
            )

    ## Options if using built-in Flask debugging server
    if "RUN_FROM_CLI" in app.config and app.config["RUN_FROM_CLI"]:
        app.config["SERVER_NAME"] = app.config["SERVER_NAME"] + ":5000"

    for key in sorted(app.config.keys()):
        if "SECRET" not in key:
            app.logger.info("{:30} = {}".format(key, app.config[key]))

    #######################################

    db = DBInterface(
        store_path=app.config["DB_STORE_PATH"],
        data_uri_base=app.config["DB_DATA_URI_BASE"],
        store_type=app.config["DB_STORE_TYPE"],
    )

    content_security_policy = None
    # content_security_policy = {
    #    "default-src": [
    #        "'self'",
    #        "cdn.jsdelivr.net",
    #    ]
    # }
    talisman = Talisman(app, content_security_policy=content_security_policy)
    app.register_blueprint(auth)
    login_manager = LoginManager()
    login_manager.login_view = "auth.login"
    login_manager.init_app(app)

    ## Modify User to have a uri
    @login_manager.user_loader
    def load_user(user_id):
        u = User(user_id)
        u.uri = os.path.join(app.config["DB_USERS_URI_BASE"], user_id)
        return u

    if app.config["PROXY_FORWARDING"]:
        app.logger.debug("Enabling ProxyFix WSGI Middleware")
        app.wsgi_app = ProxyFix(app.wsgi_app, x_for=1, x_proto=1, x_host=1, x_prefix=1)

    #######################################

    @app.route("/")
    @login_required
    def index():
        status = None
        try:
            status = request.args["status"]
        except KeyError:
            pass
        reason = None
        try:
            reason = request.args["reason"]
        except KeyError:
            pass
        return render_template(
            "index.html", db=db, urllib=urllib, status=status, reason=reason
        )

    @app.route("/feature")
    @login_required
    def features():
        feature = request.args["feature"]
        featureName = db.getLabel(feature)
        featureComment = db.getComment(feature)
        properties = db.listObservableProperties(feature)
        propInfos = []
        for prop in properties:
            propInfo = {}
            propInfo["label"] = db.getLabel(prop)
            propInfo["comment"] = db.getComment(prop)
            proptype = db.getPropertyObservableType(prop)
            match proptype:
                case db.SDTW.quantitative:
                    qk, unit = db.getPropertyQuantityKindAndUnit(prop)
                    propInfo["qk"] = db.getLabel(qk)
                    propInfo["unit"] = db.getLabel(unit)
                case db.SDTW.categorical:
                    propInfo["qk"] = "Categorical"
                    propInfo["unit"] = ""
                case GEO.point:
                    propInfo["qk"] = "Geographic point"
                    propInfo["unit"] = ""
                case _:
                    raise ValueError(
                        "Unkown property type: {proptype} for property: {prop}"
                    )
            propInfos.append(propInfo)
        featureURLEncoded = urllib.parse.quote(feature, safe="")
        return render_template(
            "feature.html",
            feature=feature,
            featureName=featureName,
            featureComment=featureComment,
            properties=propInfos,
            featureURLEncoded=featureURLEncoded,
            urllib=urllib,
        )

    @app.route("/addproperty")
    @login_required
    def addproperty():
        feature = request.args["feature"]
        featureName = db.getLabel(feature)
        featureURLEncoded = urllib.parse.quote(feature, safe="")
        status = None
        try:
            status = request.args["status"]
        except KeyError:
            pass
        reason = None
        try:
            reason = request.args["reason"]
        except KeyError:
            pass
        return render_template(
            "addproperty.html",
            feature=feature,
            featureName=featureName,
            featureURLEncoded=featureURLEncoded,
            urllib=urllib,
            quantity_kind_list=db.get_quantity_kind_list(),
            status=status,
            reason=reason,
        )

    @app.route("/selectpropertyunit")
    @login_required
    def selectpropertyunit():
        feature = request.args["feature"]
        propname = request.args["propname"]
        proptype = request.args["proptype"]
        comment = request.args["comment"]
        quantityKind = request.args["quantitykind"]
        quantityKindLabel = db.getLabel(quantityKind)
        featureURLEncoded = urllib.parse.quote(feature, safe="")
        featureName = db.getLabel(feature)
        units = db.get_units_for_quantity_kind(quantityKind)
        return render_template(
            "selectpropertyunit.html",
            feature=feature,
            featureName=featureName,
            featureURLEncoded=featureURLEncoded,
            propname=propname,
            proptype=proptype,
            comment=comment,
            quantityKind=quantityKind,
            quantityKindLabel=quantityKindLabel,
            units=units,
        )

    @app.route("/tableview")
    @login_required
    def tableview():
        feature = request.args["feature"]
        featureName = db.getLabel(feature)
        featureURLEncoded = urllib.parse.quote(feature, safe="")
        props, headings = db.getColumnHeadings(feature)
        stim_times, data, stim_comments = db.getData(feature)
        return render_template(
            "tableview.html",
            featureName=featureName,
            featureURLEncoded=featureURLEncoded,
            headings=headings,
            stim_times=stim_times,
            stim_comments=stim_comments,
            data=data,
            zip=zip,
        )

    @app.route("/enterdata")
    @login_required
    def enterdata():
        feature = request.args["feature"]
        status = None
        try:
            status = request.args["status"]
        except KeyError:
            pass
        reason = None
        try:
            reason = request.args["reason"]
        except KeyError:
            pass
        featureName = db.getLabel(feature)
        featureURLEncoded = urllib.parse.quote(feature, safe="")
        props, headings = db.getColumnHeadings(feature)
        categories = []
        proptypes = []
        for prop in props:
            categories.append(db.getCategories(prop))
            proptypes.append(db.getPropertyObservableType(prop))
        propheadings = list(
            zip([str(prop) for prop in props], headings, categories, proptypes)
        )
        return render_template(
            "enterdata.html",
            featureName=featureName,
            featureURLEncoded=featureURLEncoded,
            feature=str(feature),
            propheadings=propheadings,
            status=status,
            reason=reason,
            geopoint=GEO.point,
        )

    @app.route("/form/addfeature", methods=["post"])
    @login_required
    def form_addfeature():
        featurename = request.form["featurename"]
        comment = request.form["comment"]
        try:
            db.addNewFeature(featurename, comment)
        except Exception as e:
            return redirect(url_for("index") + "?status=error&reason=" + str(e))
        return redirect(url_for("index") + "?status=success")

    @app.route("/form/addproperty", methods=["post"])
    @login_required
    def form_addproperty():
        feature = request.form["feature"]
        propname = request.form["propname"]
        proptype = request.form["proptype"]
        comment = request.form["comment"]
        quantityKind = request.form["quantitykind"]
        unit = None
        try:
            unit = request.form["unit"]
        except KeyError:
            pass
        try:
            db.addNewObservableProperty(
                propname, proptype, comment, feature, quantityKind, unit
            )
        except Exception as e:
            return redirect(
                url_for("addproperty")
                + "?feature="
                + feature
                + "&"
                + "status=error&reason="
                + str(e)
            )
        return redirect(
            url_for("addproperty") + "?feature=" + feature + "&" + "status=success"
        )

    @app.route("/form/adddata", methods=["post"])
    @login_required
    def form_adddata():
        form = dict(request.form)
        feature = form.pop("feature")
        user_uri = current_user.uri
        try:
            db.enterData(
                feature,
                datetime.now().astimezone().replace(microsecond=0).isoformat(),
                user_uri,
                form["comment"],
                form,
            )
        except DataValidationError as e:
            return redirect(
                url_for("enterdata")
                + "?feature="
                + feature
                + "&"
                + "status=error&reason="
                + str(e)
            )
        return redirect(
            url_for("enterdata") + "?feature=" + feature + "&" + "status=success"
        )

    @app.route("/download/csv")
    @login_required
    def download_csv():
        feature = request.args["feature"]
        csv_data = db.getCSV(feature)
        return Response(csv_data, mimetype="text/csv")

    @app.route("/download/rdf/semweb")
    @login_required
    def download_rdf():
        rdf_data = db.getDataRDF()
        return Response(rdf_data, mimetype="application/trig")

    return app


if __name__ == "__main__":
    import os
    import argparse
    from werkzeug.middleware.profiler import ProfilerMiddleware

    parser = argparse.ArgumentParser(
        description="Test server for Semantic Data Taking Webapp"
    )
    parser.add_argument(
        "-p", "--profiler", action="store_true", help="Enable profiler for all calls"
    )
    args = parser.parse_args()

    os.environ["FLASK_SERVER_NAME"] = "semweb.localhost:5000"
    app = create_app()
    if args.profiler:
        app.wsgi_app = ProfilerMiddleware(app.wsgi_app)
        print("Warning: Profiling enabled")
    app.run("0.0.0.0", debug=True, port=5000, ssl_context="adhoc")  # nosec
